### ISSUE TITLE
Support for template driven UI

### ISSUE THREAD

**Issue by cajus** [qooxdoo/qooxdoo#9264]
We've some mini engine here that supports building a UI from a JSON description. It's variable enough to allow model binding, custom properties, validations, etc. It is used for model items where we do not know how they should be rendered at build time. Additionally, they make the code more readable, because all these cascades of creating new Widgets are moved into a descriptive form of the JSON file. There's a schema for the JSON file, so it can be validated on demand.

Here's a playground to see what I'm talking about: https://ferdi.naasa.net/qx/#Templates-ria

The issue want's to start some discussion whether it's useful for upstream qx - and if so, if the format is good enough to fit the needs.

**Comment by gnikolaidis** [comment]
You might also want to have a look at http://www.qookery.org/

It is in active development and it supports projects with hundreds of forms (disclaimer: I work for the company that maintains it).

**Comment by johnspackman** [comment]
I think the danger here is that there is a *lot* of scope for different styles and mechanisms of UI design and layout, it's an enormous subject and the products I can think of where UI design has become a success, the UI design tool is an enforced part of the ecosystem - EG Visual Studio, Delphi Studio, etc.

Perhaps we should showcase some of the more useful 3rd party tools on the website as part of a "Getting Started" / "Qooxdoo Orientation" guide - not just having a list of contribs or GitHub repos, but more of a curated series of articles about a select few tools and contribs that new and prospective users would find relevant.

**Comment by cajus** [comment]
@gnikolaidis yes, I know qookery.

**Comment by gnikolaidis** [comment]
@johnspackman I agree with this approach. We could contribute an article for this proposed section.

**Comment by petitben** [comment]
@cajus Looking at the example you provided I have a feeling you do a lot of eval() right ? For instance, your 'tr()' calls etc... ? I think your proposition and qookery go in the right direction but there are things that must be simplified and other ones that must be improved.

When we take a look at other JS frameworks or the next branch the approach is similar for some features and very different :

- Sencha Extjs views is based on JS definitions where each widget is aliased by an xtype (so you can use pseudo-declarative Ext.create() factories) but most of the dynamic parts are directly expressed as JS code (not strings that have to be evaluated).
- Sencha Extjs / The next branch use templates inside widget definition classes
- React/Angular etc... allow defining new virtual tags that are eventually our widgets to be rendered on the screen. QxWeb goes a bit in that direction, as well as qookery even though the approach is different (you can't insert - stop me if I'm wrong - inline HTML).

My point of view is Qooxdoo should be "inlined" more easily inside html pages and that we could declare views "à la Extjs" - except that it will be more "consistent" as the qooxdoo UI abstraction is in my point of view way better :-) ... So to sum up, the best solution is a mix between your template engine and qookery's propositions.

**Comment by cajus** [comment]
@petitben no, we're not using eval. The tr related "calls" are hard coded and needed for things like plural, message extraction by the generator, etc. One reason not to choose qookery was XML. Other reasons where special requirements to model binding for a local project, so that we abstracted parts a bit more away.

To be not misunderstood: qookery is a good piece of software, it just didn't match the requirements here. And while I think that qx would benefit from *some kind* of non programatic way to generate an interface, I just wanted to push the discussion here.

At least React/Angular aren't really comparable to qx as they are more HTML/JSX/CSS centric. The last time ExtJS evaluation is as far away as their name change/comercialization. So I don't remember. I only remember that I didn't like it at that time ;-)

**Comment by cboulanger** [comment]
As far as I can see, there are three projects that scratch a similar itch, that is, the generation of user interfaces from data instead of through procedural code:
- qookery (see #9342)
- @derrell's "qooxit" project (http://svn.code.sf.net/p/qooxdoo-contrib/code/trunk/qooxdoo-contrib/qooxit/trunk)
- @cajus's project mentioned above  (the playground link is broken, though)

Instead of reinventing the wheel three times and producing and maintaining three mutually incompatible project, wouldn't it be worthwhile to come up with a common JSON representation of qooxdoo widgets and a "builder engine" which could be integrated into core? These two elements could be the foundation of these three projects and possibly others. 

In other words, this would mean to split declarative UI generation into two parts: 
   - a builder that  works with an intermediary declarative "language" (i.e. JSON schema) to define a qooxdoo UI and is integrated into the core 
   - the other part lives in contrib libraries, and produces the JSON intermediary data

This way we would have the best of all worlds!

**Comment by oetiker** [comment]
this is my *very* simple take on on the problem https://github.com/oetiker/callbackery/blob/master/lib/CallBackery/qooxdoo/callbackery/source/class/callbackery/ui/form/Auto.js

**Comment by JosKar** [comment]
@cboulanger In theory the form parser of qookery could be replaced make it parse JSON instead of XML and with some small patches here and there it will be working. But I cannot understand why so many people propose JSON format to write a declarative definition of a UI.

Some of my initial thoughts:

- From the perspective of a new web developer seeing all this new stuff and trying to understand what is going on,  XML in my opinion would feel more natural, because mentally is closer to html. And as qookery is a thin wrapper over qooxdoo there aren't so many new staff to learn about.
- You will have to create a JSON schema in order to validate the JSON forms. I don't believe that JSON schema is equally mature as the xml schema.
- JSON forms will result into unmaintainable code.
  For example when time comes to write some logic in your form it would become messy, because
  JSON does not allow line breaks (http://json.org/ see the string definition).

Here is a simple example with a button that has a tap listener:

JSON could be like something like:
```
  "button": {
    "label": "clickme",
        "scripts": [
             { "event": "tap", "script": "if(something) { // do this } else { // do something else }" }
      ]}
   }
```
XML current way of declaring a button with a tap listener in qookery:
```
<button label="clickme">
  <script event="tap">
      if(something) {
        // do stuff
      }
      else {
        // do something else
      }
  <script>
</button>
```

What I am trying to say is that we have to use each technology where appropriate. JSON is a good technology but it doesn't seems to be the correct tool for the job here. This post is not about JSON vs XML! All I am trying to do, is to imagine how the life of a JSON form author vs a XML form author would be.

**Comment by gnikolaidis** [comment]
Further adding to @JosKar comment, XML allows us to

- validate our files against a normative schema
- use IDE features like autocomplete, error marking, etc
- import files with <xi:include>
- have  comments within files
- have line-breaks in <script> code
- be extensible using namespaces
- reuse forms for secondary purposes like documentation with XSL
- nicely nest elements with attributes like an HTML dev would expect

At Ergobyte we do not have a strong preference about the file format per se. We simply feel that reverting to JSON is a huge step backwards. Admittedly, XML is not hip in 2017, but we have yet to hear a real technical reason why we should abandon it.

**Comment by cboulanger** [comment]
@gnikolaidis Just to clarify: this is not about you abandoning XML. By all means, use XML, I use qookery too and love it. The point is that not everyone wants to work with XML, it's a matter of taste. The common problem is data- or template-driven UI design, and the step going from data (which can be expressed in JSON, XML, YAML, etc,) to qooxdoo widgets...

**Comment by cboulanger** [comment]
@gnikolaidis I think this issue is just an invitation to think about code that could be shared. If everyone wants to keep on going separate ways, that's totally fine, too. As I said, we can pull in different solutions via contribs.

**Comment by gnikolaidis** [comment]
A "matter of taste" is not a technical argument in my book. I would love to switch to another file format that gives enough (or even more) expressibility. I also do not see the benefits of JSON as an intermediary format.

**Comment by level420** [comment]
My two cents: 

I've implemented a similar aproach as @oetiker did. I'm creating via a sort of compiler an array containing UI descriptions for complex forms (see https://demodb.reanimationsregister.de) which also allows to define dependencies between form fields (widgets). This dependencies allow to enable/disable hide/show widgets and widget groups held in tabs based on values of other fields and pre set values. The source of the compiler is an XML file in conjunction with a XSD schema which allows me to create backend code, database DDL and DML and the frontend UI.

So I've choosen to have the UI description in XML and translating it to a javascript array which is handeled by a rendering pipeline.

One thing what I needed to do is to create some rendering pipeline which interrupts the widget instantiation queue in order to avoid the "a script is blocking the browser" message and to give the user some feedback while rendering large forms.

So yes I also have such a descriptive UI creation toolchain which uses XML/XSD as it's source.

My concerns in using JSON is that it does not allow function definitions whereas XML has no limitations as it allows code parts to be defined like
```xml
<field name="MYFIELD" width="180">
    <code><![CDATA[
	item.addListener("changeValue", function(e) {
	}, form);
    ]]></code>
</field>
```
which allows me to add code not covered by the descriptive UI.

**Comment by oetiker** [comment]
Note, this is NOT a discussion XML vs JSON ... 

the idea is that there is code, that creates/manipulates qx objects based on a data-structure ... 

how that data-structure is created, by parsing xml or processing a JSON file is NOT part of this ticket ... 

I think there is worth in the code that generates the qx stuff, regardless of where it is getting the information from originally.

**Comment by peuter** [comment]
As the link @cajus provided is not working, here is an example template:

```
{
    "type": "widget",
    "class": "qx.ui.container.Composite",
    "layout": "qx.ui.layout.VBox",
    "properties": {
        "categoryTitle": "tr('Country')"
    },
    "extensions": {
        "tabConfig": {
            "title": "tr('Country')",
            "icon": "favicon"
        },
        "resources": {
            "favicon": "@Ligature/globe"
        }
    },
    "children": [{
        "class": "qx.ui.container.Composite",
        "layout": "qx.ui.layout.Grid",
        "layoutConfig": {
            "spacingX": "CONST_SPACING_X",
            "spacingY": "CONST_SPACING_Y"
        },
        "extensions": {
            "layoutOptions": {
                "columnFlex": {
                    "column": 1,
                    "flex": 1
                }
            }
        },
        "children": [{
            "class": "gosa.ui.widgets.QLabelWidget",
            "buddyModelPath": "c",
            "properties": {
                "text": "tr('Name')"
            },
            "addOptions": {
                "row": 0,
                "column": 0
            }
        }, {
            "class": "gosa.ui.widgets.QComboBoxWidget",
            "modelPath": "c",
            "properties": {
                "sortBy": "value"
            },
            "addOptions": {
                "row": 0,
                "column": 1
            }
        }, {
            "class": "gosa.ui.widgets.QLabelWidget",
            "buddyModelPath": "description",
            "properties": {
                "text": "tr('Description')"
            },
            "addOptions": {
                "row": 1,
                "column": 0
            }
        }, {
            "class": "gosa.ui.widgets.QLineEditWidget",
            "modelPath": "description",
            "addOptions": {
                "row": 1,
                "column": 1
            }
        }]
    }]
}
```
It is pretty straight forward if you are used to qooxdoo. And just as it has been mentioned twice: IMHO javascript code should not be part of the template. With the given example, you can create a custom class for you needs and use that one in the template. Just my 2cts.

**Comment by hkollmann** [comment]
examples are not working.

**Comment by cboulanger** [comment]
I would leave this open since it is a conceptual issue/feature request, not so much a concrete implementation. I am still interested in the idea.
